submarine/src/main/java/org/apache/zeppelin/submarine/SubmarineInterpreter.java

```diff
@@ -0,0 +1,275 @@
+        }
+      }
+
+      String distributed = this.properties.getProperty(MACHINELEARING_DISTRIBUTED_ENABLE, "false");
+      SubmarineUtils.setAgulObjValue(context, unifyKey(MACHINELEARING_DISTRIBUTED_ENABLE),
+          distributed);
+
+      inputPath = SubmarineUtils.getAgulObjValue(context, INPUT_PATH);
+      cleanCheckpoint = SubmarineUtils.getAgulObjValue(context, CLEAN_CHECKPOINT);
+      chkPntPath = submarineJob.getJobDefaultCheckpointPath();
+      SubmarineUtils.setAgulObjValue(context, CHECKPOINT_PATH, chkPntPath);
+      psLaunchCmd = SubmarineUtils.getAgulObjValue(context, PS_LAUNCH_CMD);
+      workerLaunchCmd = SubmarineUtils.getAgulObjValue(context, WORKER_LAUNCH_CMD);
+      properties.put(INPUT_PATH, inputPath != null ? inputPath : "");
+      properties.put(CHECKPOINT_PATH, chkPntPath != null ? chkPntPath : "");
+      properties.put(PS_LAUNCH_CMD, psLaunchCmd != null ? psLaunchCmd : "");
+      properties.put(WORKER_LAUNCH_CMD, workerLaunchCmd != null ? workerLaunchCmd : "");
+
+      SubmarineCommand submarineCmd = SubmarineCommand.fromCommand(command);
+      switch (submarineCmd) {
+        case USAGE:
+          submarineJob.showUsage();
+          break;
+        case JOB_RUN:
+          if (StringUtils.equals(cleanCheckpoint, "true")) {
+            submarineJob.cleanJobDefaultCheckpointPath();
+          }
+          submarineJob.runJob();
+          break;
+        case JOB_STOP:
+          String jobName = SubmarineUtils.getJobName(userName, noteId);
+          submarineJob.deleteJob(jobName);
+          break;
+        case TENSORBOARD_RUN:
+          submarineJob.runTensorBoard();
+          break;
+        case TENSORBOARD_STOP:
+          String user = context.getAuthenticationInfo().getUser();
+          String tensorboardName = SubmarineUtils.getTensorboardName(user);
+          submarineJob.deleteJob(tensorboardName);
+          break;
+        case OLD_UI:
+          createOldGUI(context);
+          break;
+        case CLEAN_RUNTIME_CACHE:
+          submarineJob.cleanRuntimeCache();
+          break;
+        default:
+          submarineJob.onDashboard();
+          break;
+      }
+    } catch (Exception e) {
+      LOGGER.error(e.getMessage(), e);
+      return new InterpreterResult(InterpreterResult.Code.ERROR, e.getMessage());
+    }
+
+    return new InterpreterResult(InterpreterResult.Code.SUCCESS);
+  }
+
+  @Override
+  public void cancel(InterpreterContext context) {
+    SubmarineJob submarineJob = submarineContext.addOrGetSubmarineJob(properties, context);
+    String userName = context.getAuthenticationInfo().getUser();
+    String noteId = context.getNoteId();
+    String jobName = SubmarineUtils.getJobName(userName, noteId);
+    submarineJob.deleteJob(jobName);
+  }
+
+  @Override
+  public FormType getFormType() {
+    return FormType.SIMPLE;
+  }
+
+  @Override
+  public int getProgress(InterpreterContext context) {
+    return 0;
```

Comment:
can this support progress in the future?

Reply:
The submarine interpreter can display the execution status of the task more abundantly.
In addition, I discussed with Jeff that this richer UI can be packaged into components for use by other interpreters.

![JobRun](https://user-images.githubusercontent.com/3677382/54327310-9d29e600-4644-11e9-87e6-51d2421a3e75.gif)

Reply:
thats cool